### PR TITLE
4 color grayscale support

### PR DESCRIPTION
**Pull request by webphax**
Hey @samsonmking , pretty cool project!

This pull requests adds the 4 color grayscale option (configurable). I've tested it with a 4in2 display.

### CODE REVIEW

File: page.js
Comment (samsonmking) [inline page.js:21]: I think it might be a bit cleaner to try a polymorphic solution, instead of using the `color_depth` flag.  It would work the same way as the `renderPNG` method exposed by each device. See my other comment for more details

File: image.js
Comment (samsonmking) [inline image.js:4]: As I mentioned in my other comment, I think using flags to differentiate between configurations might get tricky as we add more devices. Let's just stick to exposing the individual methods for now.

- [ ] Expose each method individually

File: common.js
Comment (samsonmking) [inline common.js:2]: - [ ] Refactor to `init()` method and remove this file

**Review by samsonmking — CHANGES_REQUESTED** [review]
@webphax thanks so much for the PR! I'm very excited to integrate these changes! I've just made a few comments to help keep these changes in line with the long term direction I'd like to see for the library.

Could you also please include a new example that could be used to test the 4in2 display in gray scale? It doesn't have to be a fancy app or anything, it could just render an image or text in different shades of gray.

Please let me know if any of my comments are unclear or you have questions.

File: page.js
Comment (webphax) [reply page.js]: Done, please check 69de6b0

File: devices.js
Comment (webphax) [reply devices.js]: Totally agree. I've added both new devices and also an `init()` function for each. Please check at 69de6b0. Are you planning to look deeper into the partial refresh functionality within your 2.0 release? I had planned to start with it next.

File: image.js
Comment (webphax) [reply image.js]: Done, please check 69de6b0

**Comment by webphax** [comment]
I'm happy that you like the PR. The cool thing about this functionality is that you can directly use it with all the existing examples. It makes the whole image / text look more natural and readable. My personal project is still in development, but I'm looking forward adding it as an example as well.

File: common.js
Comment (webphax) [reply common.js]: Done, please check 69de6b0

**Review by samsonmking — CHANGES_REQUESTED** [review]
@webphax Your changes to `init` are exactly what I had in mind. I pulled the PR and ran it locally. The 4 bit grayscale looks great on my 4.2" display in the horizontal configuration - the text is so much clearer!

When I tested the vertical configuration, however, I got the output below. Probably a quick fix?

image

- [ ] Resolve vertical display issue or remove vertical configuration

File: devices.js
Comment (samsonmking) [reply devices.js]: At the minimum, I'm thinking the 2.0 release will include breaking changes to the device API. The ability to get a device with a factory, so that parameters (orientation, color mode, etc) can be passed in and the drivers can be lazy loaded.

Feel free to go ahead and start taking a look at the partial refresh functionality! We could release that before or after the 2.0 API changes. If the partial refresh ends up requiring breaking API changes, then we might want to coordinate on timing a little bit so it's only one breaking release.

**Comment by webphax** [comment]
> @webphax Your changes to `init` are exactly what I had in mind. I pulled the PR and ran it locally. The 4 bit grayscale looks great on my 4.2" display in the horizontal configuration - the text is so much clearer!
> 
> When I tested the vertical configuration, however, I got the output below. Probably a quick fix?
> 
> image
> 
>     * [ ]  Resolve vertical display issue or remove vertical configuration

Should be fixed with c9314d3. It was using the BW buffer instead of the 4Grey buffer.

**Comment by webphax** [comment]
Oh no...wrong button

**Comment by samsonmking** [comment]
@webphax I just pulled and tested again with your latest commit. It looks different, but the vertical config still looks a bit off to me. 

Can you double check it's working on your end?
image

**Review by samsonmking — CHANGES_REQUESTED** [review]
@webphax I pulled the latest and tested again. It looks different, but still a little off.

Can you double check it works on your end? Thanks for the quick replies and let me know if there's anything I can do to help!

image

**Comment by samsonmking** [comment]
@webphax this seems to work for me:
```javascript
function convertPNGto1Bit4GreyRotated(pngBytes) {
    const reader = new PNGReader(pngBytes);
    return new Promise((resolve, reject) => {
        reader.parse((err, png) => {
            if (err) {
                return reject(err);
            }
            const height = png.getHeight();
            const width = png.getWidth();
            const devHeight = width;
            const devWidth = height;
            const outBuffer = allocBuffer_4(devWidth, devHeight);
            var i = 0;
            for (let x = 0; x < width; x++) {
                for (let y = 0; y < height; y++) {
                    const outX = y;
                    const outY = devHeight - x - 1;
                    i++;
                    if (i % 4 == 0) {
                        out_index = Math.floor((outX + outY * devWidth) / 4)
                        outBuffer[out_index] = (
                            (RGBAToHex(png.getPixel(x, y - 3)) & 0xc0) |
                            (RGBAToHex(png.getPixel(x, y - 2)) & 0xc0) >> 2 |
                            (RGBAToHex(png.getPixel(x, y - 1)) & 0xc0) >> 4 |
                            (RGBAToHex(png.getPixel(x, y)) & 0xc0) >> 6
                        )
                    }
                }
            }
            resolve(outBuffer);
        });
    });
}
```

One more question, though. I was looking through the Waveshare example python implementations. It seems they use the python library PIL to convert the image to 8bit grayscale, before creating the grayscale buffer. See the documentation, [here](https://pillow.readthedocs.io/en/stable/handbook/concepts.html#concept-modes).
```python
def getbuffer_4Gray(self, image):
        buf = [0xFF] * (int(self.width / 4) * self.height)
        image_monocolor = image.convert('L')
        ...
```

Should we do something similar? It's probably not noticeable with the current examples, because the webpage only uses black and white colors. But it might be nice for users of epaper.js to not to have to worry about picking the correct colors in their HTML / CSS. Looks like the library [Sharp](https://sharp.pixelplumbing.com/api-colour#greyscale) could provide this functionality pretty easily. Would you be interested in adding this to your implementation?

**Comment by webphax** [comment]
I've updated the 'function convertPNGto1Bit4GreyRotated(pngBytes)' according to your changes in commit 724fd27. I thought adding grayscale() via sharp would be a quick fix, but I'm running into some ugly errors. I'll need to evaluate further and will probably commit an update later today.

**Comment by webphax** [comment]
@samsonmking it was a problem with an incorrect sharp installation. A clean reinstall fixed it. As suggested it now converts the png image to greyscale first.

**Comment by webphax** [comment]
I am not quite sure but Git forced me to resolve merge conflicts and create a new commit (see b03b665). With this even untouched files changed. Could you please have a look at it and exclude it if necessary?

**Comment by samsonmking** [comment]
@webphax sorry for the delay. I just pulled and tested your changes - looks good!

Would you mind squashing your commits down to a single commit? I should have provided more detailed contribution instructions so you'd have known ahead of time. Once the commits are squashed I'll merge and release!

**Review by samsonmking — APPROVED** [review]

**Comment by webphax** [comment]
@samsonmking great that you had time to check the changes! I am not really familiar with the GitHub PR concept, but I assume that you shoud have a ["Squash and merge"-option](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/about-pull-request-merges#squash-and-merge-your-pull-request-commits) when mergin the PR changes with your master branch. As I don't have write access to your repository I could 
a) squash and merge with 'my' master branch or 
b) create a new branch and a new PR only containing the squashed commit

**Comment by samsonmking** [comment]
@webphax good suggestion for the squash and merge option. The setup I use at work requires me to do it manually, so I'd forgotten that was an option.

I've released this as 1.3.0! It's great to have others contributing to this project!

**Pull request merged**
